### PR TITLE
Initial public draft for TotW 109

### PR DESCRIPTION
**Pull request by tituswinters**

### CODE REVIEW

File: _posts/2017-11-23-totw-109.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+void F(int);                     // 1: declaration of F(int)
+void F(const int);               // 2: re-declaration of F(int)
+void F(int) { /* ... */ }        // 3: definition of F(int)
+void F(const int) { /* ... */ }  // 4: error: re-definition of F(int)
+```
+
+The first two lines are function *declarations*. A function *declaration* tells
+the compiler the function's signature and return type. In the above example, the
+function's signature is `F(int)`. The constness of the function's parameter type
+is ignored, so both declarations are equivalent ([over.load][CppStandard]).
```
Comment (manshreck) [inline _posts/2017-11-23-totw-109.md]: Need precise link to the C++ standard, if possible, or unlinkify if not. (This only links to the base page and loses the over.load link.)

Also, my strong personal preference is to always name the heading being linked to. (In this case "Overloadable declarations".)

So I'd write this as.

See ["Overloadable declarations"](http://eel.is/c++draft/over.load) in the C++ standard.

File: _posts/2017-11-23-totw-109.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+declarations are allowed, but only a single definition is permitted.
+
+Even though the definitions on lines 3 and 4 *declare* and *define* the same
+function, there is a difference within their function bodies due to the way they
+are declared. From the definition on line 3, the type of the function-parameter
+variable within the function will be `int` (i.e., non-const). On the other hand,
+the definition on line 4 will produce a function-parameter variable within the
+function whose type is `const int`.
+
+### Meaningful `const` in function declarations
```
Comment (manshreck) [inline _posts/2017-11-23-totw-109.md]: Initial Caps Headings

File: _posts/2017-11-23-totw-109.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+Not all `const` qualifications in function declarations are ignored. To quote
+from [over.load][CppStandard] in the C++ standard (emphasis added):
```
Comment (manshreck) [inline _posts/2017-11-23-totw-109.md]: "Overloadable declarations" ([over.load]) in the C++ standard.

File: _posts/2017-11-23-totw-109.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+> `const` type-specifiers **buried within a parameter type specification** are
+> significant and can be used to distinguish overloaded function declarations
+
+The following are examples where `const` is significant and not ignored.
```
Comment (manshreck) [inline _posts/2017-11-23-totw-109.md]: Nit:

ignored:

File: _posts/2017-11-23-totw-109.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+void F(std::unique_ptr<const int> x);  // 3
+void F(int* x);                        // 4
+```
+
+In the above examples, the `x` parameter itself is never declared `const`. Each
+of the above functions accepts a parameter named `x` of a different type, thus
+forming a valid overload set. Line 1 declares a function that accepts a "pointer
+to an `int` that is `const`". Line 2 declares a function that accepts a
+"reference to an `int` that is `const`". And line 3 declares a function that
+accepts a "unique_ptr of `int`s that are `const`". All of these uses of `const`
```
Comment (manshreck) [inline _posts/2017-11-23-totw-109.md]: of an \`int`\ that is (?)


(The way it's phrased suggests it accepts an array of ints ...)

File: _posts/2017-11-23-totw-109.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+1.  Never use top-level `const` on function parameters in *declarations* that
+    are not definitions (and be careful not to copy/paste a meaningless
+    `const`). It is meaningless and ignored by the compiler, it is visual noise,
+    and it could mislead readers.
+2.  Do use top-level `const` on function parameters in *definitions* at your (or
+    your team's) discretion. You might follow the same rationale as you would
+    for when to declare a function-local variable `const`.
+
+[CppStandard]: http://eel.is/c++draft/
```
Comment (manshreck) [inline _posts/2017-11-23-totw-109.md]: Remove if fixing links as indicated above.

File: _posts/2017-11-23-totw-109.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/109
+type: markdown
+order: "109"
+---
+
+Originally posted as totw/109 on 2016-01-14
+
+By Greg Miller (jgm@google.com)
```
[the remark this answers is not in the document]
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: done

File: _posts/2017-11-23-totw-109.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+This document will explain when `const` is meaningful in function declarations,
+and when it is meaningless and best omitted. But first, let us briefly explain
+what is meant by the terms *declaration* and *definition*.
+
+Consider the following code.
```
[the remark this answers is not in the document]
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: done

File: _posts/2017-11-23-totw-109.md (hunk 1 above)
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: done

File: _posts/2017-11-23-totw-109.md (hunk 2 above)
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: done

File: _posts/2017-11-23-totw-109.md (hunk 3 above)
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: done

File: _posts/2017-11-23-totw-109.md (hunk 4 above)
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: done

File: _posts/2017-11-23-totw-109.md (hunk 5 above)
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: I don't follow

File: _posts/2017-11-23-totw-109.md (hunk 6 above)
Comment (tituswinters) [reply _posts/2017-11-23-totw-109.md]: Done

File: _posts/2017-11-23-totw-109.md (hunk 5 above)
Comment (manshreck) [reply _posts/2017-11-23-totw-109.md]: The original comment is plural

declares a function that accepts a "unique_ptr of `int`s that are `const`"

vs.

declares a function that accepts a "unique_ptr of an `int`s that is `const`"

**Comment by tituswinters** [comment]
done, ptal

**Pull request merged**